lib/msprime.c

```diff
@@ -5881,159 +5856,61 @@ beta_generation_rate_to_model_rate(simulation_model_t *model, double rate)
-    struct beta_integral_params params = {num_ancestors, alpha};
-
-    if (w == NULL) {
-        ret = MSP_ERR_NO_MEMORY;
-        goto out;
-    }
-    F.function = &beta_integrand;
-    F.params = &params;
-    ret = gsl_integration_qags(&F, 0, truncation_point, epsabs, epsrel,
-            workspace_size, w, result, &err);
-    if (ret != 0) {
-        /* It's ugly, but it should only happen while tuning models so we output
-         * the GSL error to stderr */
-        fprintf(stderr,  "GSL error: %s\n", gsl_strerror(ret));
-        ret = MSP_ERR_INTEGRATION_FAILED;
-    }
-out:
-    return ret;
-}
-
-int
-msp_beta_compute_coalescence_rate(msp_t *self, unsigned int num_ancestors, double *result)
-{
-    int ret = 0;
-    double alpha = self->model.params.beta_coalescent.alpha;
-
-    *result = 1;
-    if (num_ancestors > 2) {
-        ret = msp_compute_beta_integral(self, num_ancestors, alpha, result);
-    }
-    return ret;
+    return rate / (4 * model->params.beta_coalescent.scalar);
 }
 
 static double
 msp_beta_get_common_ancestor_waiting_time(msp_t *self, population_id_t pop_id,
         label_id_t label)
 {
-    int ret = 0;
-    double result, lambda;
     population_t *pop = &self->populations[pop_id];
     unsigned int n = (unsigned int) avl_count(&pop->ancestors[label]);
-
-    ret = msp_beta_compute_coalescence_rate(self, n, &lambda);
-    if (ret != 0) {
-        assert(ret < 0);
-        /* An error occured, and we signal this back using a negative waiting time */
-        result = ret;
-    } else {
-        result = msp_get_common_ancestor_waiting_time_from_rate(self, pop, lambda);
-    }
+    /* Factor of 4 because only 1/4 of binary events result in a merger due to
+     * diploidy, and 2 for consistency with the hudson model */
+    double lambda = 8 * gsl_sf_choose(n, 2);
+    double result = msp_get_common_ancestor_waiting_time_from_rate(self, pop, lambda);
     return result;
 }
 
 int MSP_WARN_UNUSED
-msp_multi_merger_common_ancestor_event(msp_t *self, double x,
-    avl_tree_t *ancestors, avl_tree_t *Q)
+msp_multi_merger_common_ancestor_event(msp_t *self, avl_tree_t *ancestors,
+                                       avl_tree_t *Q, uint32_t k)
 {
     int ret = 0;
-    uint32_t j, k, i;
+    uint32_t j, i, l;
     avl_node_t  *node, *q_node;
     segment_t *u;
+    uint32_t pot_size;
+    uint32_t cumul_pot_size = 0;
+
     /* In the multiple merger regime we have four different 'pots' that
      * lineages get assigned to, where all lineages in a given pot are merged into
      * a common ancestor.
      */
-    k = gsl_ran_binomial(self->rng, x, avl_count(ancestors));
-    for (i = 0; i < k; i++){
-        j = (uint32_t) gsl_rng_uniform_int(self->rng, avl_count(ancestors));
-        node = avl_at(ancestors, j);
-        assert(node != NULL);
+    for (i = 0; i < 4; i++) {
+        pot_size = gsl_ran_binomial(self->rng, 1.0 / (4 - (int)i),
```

Comment:
Simpler/less likely to have obscure problems: ``1.0 / (4.0 - i)``. The ``i`` should be automatically cast to double then.

Reply:
Thanks! The rules around automatic casting give me endless headaches, especially when dealing with unsigned quantities.